File: python-playwright-enterprise-framework/pages/pim_page.py
import allure
from playwright.sync_api import expect

from utils.logger import get_logger


class PimPage:
    def __init__(self, page):
        self.page = page
        self.logger = get_logger(self.__class__.__name__)

        self.pim_menu = page.get_by_role("link", name="PIM")
        self.page_header = page.locator("h6")
        self.page_heading = page.locator("h6.orangehrm-main-title")

        self.add_button = page.get_by_role("button", name="Add")
        self.first_name_input = page.locator("input[name='firstName']")
        self.middle_name_input = page.locator("input[name='middleName']")
        self.last_name_input = page.locator("input[name='lastName']")
        self.employee_id_input = page.locator(".oxd-grid-2 input.oxd-input").first
        self.save_button = page.get_by_role("button", name="Save")

        # self.employee_name_input = page.locator(
        #     "label:has-text('Employee Name')"
        # ).locator("xpath=../../following-sibling::div//input")
        self.employee_name_input = page.locator("div.oxd-grid-item:nth-child(1) > div:nth-child(1) > div:nth-child(2) > div:nth-child(1) > div:nth-child(1) > input:nth-child(2)")
        self.search_button = page.get_by_role("button", name="Search")
        self.reset_button = page.get_by_role("button", name="Reset")

        self.records_found_label = page.locator(".orangehrm-horizontal-padding span")
        self.result_rows = page.locator(".oxd-table-card")
        self.employee_details_header = page.locator("h6").filter(
            has_text="Personal Details"
        )

    @allure.step("Navigate to PIM module")
    def navigate_to_pim(self):
        self.logger.info("Navigating to PIM module")

        self.pim_menu.click()
        expect(self.page_header).to_have_text("PIM")

    @allure.step("Click Add Employee button")
    def click_add_employee(self):
        self.logger.info("Clicking Add Employee button")

        self.add_button.click()
        expect(self.page_heading).to_have_text("Add Employee")

    @allure.step("Enter employee first name")
    def enter_first_name(self, first_name):
        self.logger.info("Entering employee first name")

        self.first_name_input.fill(first_name)

    @allure.step("Enter employee middle name")
    def enter_middle_name(self, middle_name):
        self.logger.info("Entering employee middle name")

        self.middle_name_input.fill(middle_name)

    @allure.step("Enter employee last name")
    def enter_last_name(self, last_name):
        self.logger.info("Entering employee last name")

        self.last_name_input.fill(last_name)

    @allure.step("Enter employee ID")
    def enter_employee_id(self, employee_id):
        self.logger.info("Entering employee ID")

        self.employee_id_input.fill(employee_id)

    @allure.step("Save employee")
    def click_save(self):
        self.logger.info("Saving employee")

        self.save_button.click()

    @allure.step("Add new employee")
    def add_employee(self, first_name, middle_name, last_name, employee_id):
        self.click_add_employee()
        self.enter_first_name(first_name)
        self.enter_middle_name(middle_name)
        self.enter_last_name(last_name)
        self.enter_employee_id(employee_id)
        self.click_save()

    @allure.step("Verify employee personal details page is loaded")
    def verify_employee_details_loaded(self):
        self.logger.info("Verifying employee details page is loaded")
        self.employee_details_header.wait_for(state="visible", timeout=20000)
        expect(self.employee_details_header).to_be_visible()

    @allure.step("Search employee by name")
    def search_employee_by_name(self, employee_name):
        self.logger.info("Searching employee by name: %s", employee_name)

        self.employee_name_input.fill(employee_name)
        self.page.keyboard.press("ArrowDown")
        self.page.keyboard.press("Enter")
        self.search_button.click()

    @allure.step("Verify PIM results are displayed")
    def verify_results_displayed(self):
        expect(self.records_found_label).to_be_visible()

        records_text = self.records_found_label.inner_text()
        
        assert "Record Found" in records_text or "Records Found" in records_text

    @allure.step("Verify PIM table contains text")
    def verify_table_contains_text(self, expected_text):
        expect(self.result_rows.first).to_be_visible()

        table_text = self.result_rows.first.inner_text()

        assert expected_text in table_text

    @allure.step("Click PIM reset button")
    def click_reset(self):
        self.logger.info("Clicking PIM reset button")

        self.reset_button.click()

File: javascript-playwright-enterprise-framework/pages/pim_page.js
import { expect } from "@playwright/test";
import { getLogger } from "../utils/logger.js";

export class PimPage {
  constructor(page) {
    this.page = page;
    this.logger = getLogger(this.constructor.name);

    this.pimMenu = page.getByRole("link", { name: "PIM" });
    this.pageHeader = page.locator("h6");
    this.pageHeading = page.locator("h6.orangehrm-main-title");

    this.addButton = page.getByRole("button", { name: "Add" });
    this.firstNameInput = page.locator("input[name='firstName']");
    this.middleNameInput = page.locator("input[name='middleName']");
    this.lastNameInput = page.locator("input[name='lastName']");
    this.employeeIdInput = page.locator(".oxd-grid-2 input.oxd-input").first();
    this.saveButton = page.getByRole("button", { name: "Save" });

    this.employeeNameInput = page.locator("div.oxd-grid-item:nth-child(1) > div:nth-child(1) > div:nth-child(2) > div:nth-child(1) > div:nth-child(1) > input:nth-child(2)");
    this.searchButton = page.getByRole("button", { name: "Search" });
    this.resetButton = page.getByRole("button", { name: "Reset" });

    this.recordsFoundLabel = page.locator(".orangehrm-horizontal-padding span");
    this.resultRows = page.locator(".oxd-table-card");
    this.employeeDetailsHeader = page.locator("h6").filter({ hasText: "Personal Details" });
  }

  async navigateToPim() {
    this.logger.info("Navigating to PIM module");
    await this.pimMenu.click();
    await expect(this.pageHeader).toHaveText("PIM");
  }

  async clickAddEmployee() {
    this.logger.info("Clicking Add Employee button");
    await this.addButton.click();
    await expect(this.pageHeading).toHaveText("Add Employee");
  }

  async enterFirstName(firstName) {
    this.logger.info("Entering employee first name");
    await this.firstNameInput.fill(firstName);
  }

  async enterMiddleName(middleName) {
    this.logger.info("Entering employee middle name");
    await this.middleNameInput.fill(middleName);
  }

  async enterLastName(lastName) {
    this.logger.info("Entering employee last name");
    await this.lastNameInput.fill(lastName);
  }

  async enterEmployeeId(employeeId) {
    this.logger.info("Entering employee ID");
    await this.employeeIdInput.fill(employeeId);
  }

  async clickSave() {
    this.logger.info("Saving employee");
    await this.saveButton.click();
  }

  async addEmployee(firstName, middleName, lastName, employeeId) {
    await this.clickAddEmployee();
    await this.enterFirstName(firstName);
    await this.enterMiddleName(middleName);
    await this.enterLastName(lastName);
    await this.enterEmployeeId(employeeId);
    await this.clickSave();
  }

  async verifyEmployeeDetailsLoaded() {
    this.logger.info("Verifying employee details page is loaded");
    await this.employeeDetailsHeader.waitFor({ state: "visible", timeout: 20000 });
    await expect(this.employeeDetailsHeader).toBeVisible();
  }

  async searchEmployeeByName(employeeName) {
    this.logger.info("Searching employee by name: %s", employeeName);
    await this.employeeNameInput.fill(employeeName);
    await this.page.keyboard.press("ArrowDown");
    await this.page.keyboard.press("Enter");
    await this.searchButton.click();
  }

  async verifyResultsDisplayed() {
    await expect(this.recordsFoundLabel).toBeVisible();
    const recordsText = await this.recordsFoundLabel.innerText();
    expect(recordsText.includes("Record Found") || recordsText.includes("Records Found")).toBeTruthy();
  }

  async verifyTableContainsText(expectedText) {
    await expect(this.resultRows.first()).toBeVisible();
    const tableText = await this.resultRows.first().innerText();
    expect(tableText).toContain(expectedText);
  }

  async clickReset() {
    this.logger.info("Clicking PIM reset button");
    await this.resetButton.click();
  }
}
